File: 03a/server_b/parser/parser.go
package parser

import (
	"encoding/csv"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// BaseDir points to the root directory (Mandatory 1/)
var BaseDir string

func init() {
	// Get the directory of parser.go
	dir, err := os.Getwd()
	if err != nil {
		panic(fmt.Sprintf("failed to get working directory: %v", err))
	}
	// Navigate up to Mandatory_1/ and then to data/
	// From 03a/server_b/, we need to go up two directories: server_b/ -> 03a/ -> Mandatory_1/
	BaseDir = filepath.Join(dir, "..", "..", "data")
	// Resolve to absolute path to avoid relative path issues
	BaseDir, err = filepath.Abs(BaseDir)
	if err != nil {
		panic(fmt.Sprintf("failed to resolve absolute path for BaseDir: %v", err))
	}
}

// ParseFile parses a file based on its type and set name
func ParseFile(setName, fileType string) (map[string]interface{}, error) {
	// File path: data/{set_name}/{set_name}.{file_type}
	filePath := filepath.Join(BaseDir, setName, fmt.Sprintf("%s.%s", setName, fileType))

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return nil, fmt.Errorf("file %s not found", filePath)
	}

	switch fileType {
	case "txt":
		return parseText(filePath)
	case "xml":
		return parseXML(filePath)
	case "yaml":
		return parseYAML(filePath)
	case "json":
		return parseJSON(filePath)
	case "csv":
		return parseCSV(filePath)
	default:
		return nil, fmt.Errorf("unsupported file type: %s", fileType)
	}
}

func parseText(filePath string) (map[string]interface{}, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	data := make(map[string]interface{})
	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		if strings.Contains(line, ": ") {
			parts := strings.SplitN(line, ": ", 2)
			data[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
		}
	}
	return data, nil
}

func parseXML(filePath string) (map[string]interface{}, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	// Define a generic struct to handle XML elements
	type Element struct {
		XMLName xml.Name
		Value   string `xml:",chardata"`
	}

	type Root struct {
		Children []Element `xml:",any"`
	}

	var root Root
	if err := xml.Unmarshal(content, &root); err != nil {
		return nil, err
	}

	data := make(map[string]interface{})
	for _, child := range root.Children {
		data[child.XMLName.Local] = child.Value
	}
	return data, nil
}

func parseYAML(filePath string) (map[string]interface{}, error) {
	data := make(map[string]interface{})
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	if err := yaml.Unmarshal(content, &data); err != nil {
		return nil, err
	}
	return data, nil
}

func parseJSON(filePath string) (map[string]interface{}, error) {
	data := make(map[string]interface{})
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	if err := json.Unmarshal(content, &data); err != nil {
		return nil, err
	}
	return data, nil
}

func parseCSV(filePath string) (map[string]interface{}, error) {
	data := make(map[string]interface{})
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) < 2 {
		return nil, fmt.Errorf("CSV file must have at least one header row and one data row")
	}

	headers := records[0]
	values := records[1] // Take the first data row only
	if len(values) != len(headers) {
		return nil, fmt.Errorf("row length mismatch in CSV file")
	}
	for i, header := range headers {
		data[header] = values[i]
	}
	return data, nil
}

File: 03a/server_a/src/parser.py
from pathlib import Path
import yaml
import xml.etree.ElementTree as ET
import json
import csv
import logging

# Set up logging
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

# Define the base directory (should point to Mandatory_1/)
BASE_DIR = Path(__file__).resolve().parent.parent.parent.parent
DATA_DIR = BASE_DIR / "data"

def parse_file(set_name: str, file_type: str) -> dict:
    file_path = DATA_DIR / set_name / f"{set_name}.{file_type}"
    
    logger.debug(f"Attempting to access file: {file_path}")
    if not file_path.exists():
        raise FileNotFoundError(f"File not found: {file_path}")

    if file_type == "txt":
        return parse_text(file_path)
    elif file_type == "xml":
        return parse_xml(file_path)
    elif file_type == "yaml":
        return parse_yaml(file_path)
    elif file_type == "json":
        return parse_json(file_path)
    elif file_type == "csv":
        return parse_csv(file_path)
    else:
        raise ValueError(f"Unsupported file type: {file_type}")

def parse_text(file_path: Path) -> dict:
    result = {}
    with open(file_path, "r") as f:
        for line in f:
            key, value = line.strip().split(": ", 1)
            result[key] = value
    return result

def parse_xml(file_path: Path) -> dict:
    tree = ET.parse(file_path)
    root = tree.getroot()
    result = {}
    for child in root:
        result[child.tag] = child.text
    return result

def parse_yaml(file_path: Path) -> dict:
    with open(file_path, "r") as f:
        return yaml.safe_load(f)

def parse_json(file_path: Path) -> dict:
    with open(file_path, "r") as f:
        return json.load(f)

def parse_csv(file_path: Path) -> dict:
    result = {}
    with open(file_path, "r") as f:
        reader = csv.DictReader(f)
        for row in reader:
            return row  # Assuming one row of data
    return result